RFD 33-37: Windows desktop access

awly opened the pull request:
This is a sequence of RFDs for implementing Desktop Access for Windows.

RFD 33 defines the high-level plumbing for any target OS (not just
Windows), including the client web app.
RFD 34 defines the Windows integration pieces.
RFD 35 defines the authentication mechanism used in Windows
(RFD 36 was already taken)
RFD 37 defines the OS-agnostic wire protocol used by the client web app.

Commit:
RFD 33-37: Windows desktop access

This is a sequence of RFDs for implementing Desktop Access for Windows.

RFD 33 defines the high-level plumbing for any target OS (not just
Windows), including the client web app.
RFD 34 defines the Windows integration pieces.
RFD 35 defines the authentication mechanism used in Windows
(RFD 36 was already taken)
RFD 37 defines the OS-agnostic wire protocol used by the client web app.

klizhentas commented on rfd/0034-desktop-access-windows.md:
```diff
@@ -0,0 +1,268 @@
+- list of Active Directory-enrolled hosts obtained from AD via LDAPS (LDAP over
+  SSL)
+  - LDAP library: https://pkg.go.dev/github.com/go-ldap/ldap/v3
+- local host, when running on a Windows machine in agent mode
```
@r0mant is currently refactoring discovery for application access and DB access, to support dynamic registration of the apps and databases using API as well. This will enable both auto-discovery. Take a look [here](https://github.com/gravitational/teleport/pull/6787)

awly replied:
Good idea. It should be easy to implement by exposing `WindowsDesktop` objects to `tctl create/rm`.
There are some questions about routing, but we can make simplifying assumptions like: if you dynamically register a windows machine, it has to be reachable from any `windows_desktop_service` running in gateway mode.

klizhentas commented on rfd/0034-desktop-access-windows.md:
```diff
@@ -0,0 +1,268 @@
+  - LDAP library: https://pkg.go.dev/github.com/go-ldap/ldap/v3
+- local host, when running on a Windows machine in agent mode
+
+### Concurrent sessions
```
just a note, we should support locks in phase 2 [locks](https://github.com/gravitational/teleport/pull/7286)

awly replied:
Yep, the plan is to support all the same controls that we do for other Teleport sessions.
Clarified.

klizhentas commented on rfd/0034-desktop-access-windows.md:
```diff
@@ -0,0 +1,268 @@
+  # (optional) static_hosts is a list of hostnames to register as
+  # WindowsDesktop objects in Teleport.
+  # These are usually non-AD hosts.
+  static_hosts:
```
may be just hosts? Also take a look at Roman's RFD. Lots of times users would want to use terraform provider to specify hosts, but AD based discovery is the most important part, I agree

awly replied:
renamed to `hosts`
and yes, dynamic registration will also be supported

klizhentas commented on rfd/0034-desktop-access-windows.md:
````diff
@@ -0,0 +1,268 @@
+```yaml
+windows_desktop_service:
+  enabled: yes # default false
+  listen_addr: 0.0.0.0:3028
````
Can we take advantage of Marek's TLS SNI work and do not even introduce a new port? Would it slow us down too much? @r0mant @smallinsky what release are planning SNI merged in?

awly replied:
Yep, shouldn't be a problem as soon as the SNI changes land.

klizhentas commented on rfd/0037-desktop-access-protocol.md:
````diff
@@ -0,0 +1,191 @@
+Typical sequence of messages in a desktop session:
+
+```
++--------+                                +--------+
````
I had success using improbable engineering grpc over websockets transport:

https://github.com/improbable-eng/grpc-web

It supports streaming, typescript and could save you some time

awly replied:
`grpc-web` right now does not support client-side streaming: https://github.com/improbable-eng/grpc-web#client-side-streaming
so for sending user input (mouse movement/clicks, keyboard) we'd need to use unary RPCs
I'm worried that this will add too much overhead and make the input even more sluggish than it is over a websocket.

I think the protocol described here is going to be easy to implement.
We can always swap it for gRPC later on, for maintainability - it'll simply mean adding a new proxy endpoint for the gRPC version of the protocol.

klizhentas reviewed (commented):
Have a couple of high level comments, and a suggestion on giving improbable's grpc - websocket a look, otherwise LGTM.

Commit:
Small tweaks based on review feedback

awly commented on rfd/0034-desktop-access-windows.md:
```diff
@@ -0,0 +1,272 @@
+
+Teleport will record active desktop sessions into the simplest possible format
+and then transcode them to MP4 or WebM asynchronously. Transcoding can be done
+using `ffmpeg` if we don't find a solid encoder in Go or Rust.
```
I don't want to lock users into using our proprietary "video" player for vieweing recordings.
Just in case they want to export recordings to a SIEM or even just back them up.

For the initial version, we can just not show any recordings. Only audit logs.

As for metadata - it will be a bitmap (literally a PNG image) with a timestamp.
There's no frame rate - new bitmaps appear when something changes on the screen, there can be minutes of gaps between "frames".

alex-kovoy replied:
The wasm section was about exporting a session (saving on disk) on demand. This way we do not need to store/process videos on the backend or build infra for that.

Users will have to use Teleport Web UI to view it regardless, this is because they have to be logged in to view it. Plugins can handle the video encoding during data export.

alex-kovoy replied:
Exporting can also be done on the server side (on demand as well).

alex-kovoy replied:
If this is an PNG image then we should be able to control the quality of it when storing/transmitting. 

The rdp frame rate is 30FPS, it might be possible to tweak/change it when storing too.

awly replied:
Ahh, sorry, misinterpreted your comment.
Yeah, using wasm-based encoding on export in the browser seems much easier. I'll add that.
It's possible that in the future someone will want to export recordings straight from the backend (not via UI), same as they do with audit logs. But I won't over-engineer until necessary.

And yes, I'll optimize the video stream for maximum compression.

awly replied:
Moved to RFD 33, since we'll record our custom wire protocol and not RDP.

timothyb89 commented on rfd/0035-desktop-access-windows-authn.md:
```diff
@@ -0,0 +1,119 @@
+against a trusted Certificate Authority. More information at
+https://docs.microsoft.com/en-us/troubleshoot/windows-server/windows-security/enabling-smart-card-logon-third-party-certification-authorities
+
+Smart cards are typically hardware devices, but a driver can be
+implemented to emulate them in software
+([example](http://frankmorgner.github.io/vsmartcard/virtualsmartcard/README.html)).
+Since this is the only known way for us to support cert-based authn, Teleport
+will implement a smart card emulator to authenticate over RDP. This emulator
+will use Teleport TLS certs and expose the User TLS CA for import into the
+Windows trust store.
```
Would this be a proper Windows device driver? If so, I wonder if we could alternatively implement our smart card "driver" directly inside the RDP client. Though I'm not sure what's better: a forked RDP library or a custom Windows driver...

awly replied:
this will be an emulated driver implemented in the RDP client, since the client (`windows_desktop_service`) will likely run on a Linux machine.
RDP server should establish a virtual channel back to the client to send driver IOCTLs

timothyb89 replied:
Ah, makes sense, thanks!

Commit:
Move session recording section to RFD 33

Also change to using wasm-based ffmpeg on the client to export a video
recording.

Commit:
Merge branch 'master' into rfd/003x-desktop-access

zmb3 commented on rfd/0033-desktop-access.md:
```diff
@@ -0,0 +1,176 @@
+  against
+- `desktop_clipboard` - option for allowing copy/paste to/from the desktop
+  host; default is `true`
+- `record_desktop_session` - option for enabling/disabling recording of
```
Would there be any other way to enable/disable session recording outside of the role configuration? Will this model allow us to support configurations such as "always record sessions in the prod environment, but do not record in the dev environment?"

awly replied:
That's a good question.
You can achieve the example configuration you gave by creating prod- and dev-specific roles, which seems like a good practice anyway.

We could add some additional recording configuration at the `windows_desktop_service` level, but it would make the rules of when recording happens complex. Configuring teleport is already challenging for most users.

I'd like to stick with our current RBAC-centric model unless there's a high demand for something more flexible

zmb3 commented on rfd/0033-desktop-access.md:
```diff
@@ -0,0 +1,176 @@
+video format like MP4 or WebM.
+
+To convert our storage format to the video format, we will use [wasm-based
+ffmpeg](https://github.com/ffmpegwasm/ffmpeg.wasm) in the browser. Conversion
```
Any experience with how ffmpeg performs in the browser? I imagine an older workstation with a lot of browser tabs open might get bogged down trying to transcode video.

awly replied:
I haven't benchmarked it, but it will definitely be slow and resource-intensive.
However, this will be a relatively rare process - most users can watch the replay on our web UI without exporting and transcoding the video.

zmb3 commented on rfd/0034-desktop-access-windows.md:
```diff
@@ -0,0 +1,256 @@
+It can also talk to `localhost` RDP service, if installed on a Windows machine
+in agent mode (described below).
+
+`windows_desktop_service` also discovers all available Windows hosts from
```
I see below under host discovery that this is not the only option and a hard-coded list of hosts is also supported. It might help to clarify that here. On first read I took this to mean that all hosts on the domain would be automatically exposed to teleport (which may or may not be desirable)

awly replied:
Clarified in https://github.com/gravitational/teleport/pull/7841, please review

ikkisoft commented on rfd/0033-desktop-access.md:
```diff
@@ -0,0 +1,176 @@
+Role definitions for RBAC have a few new fields for Desktop Access:
+- `${OS}_desktop_logins` - list of desktop login names allowed/denied for
+  desktop hosts with a given OS
+- `${OS}_desktop_labels` - list of labels to match the target desktop hosts
```
How is matching between labels and actual hosts done?

awly replied:
Same way as for SSH nodes and all other Teleport resources: https://goteleport.com/docs/access-controls/getting-started/

Resource labels are a set of `key: value` pairs.
Each Teleport role grants some set of corresponding `key: list of values` pairs.
When a user (via their roles) has a superset of all the `windows_desktop_labels` that a target desktop has, they are granted access.
When a resource has at least one label that the user doesn't have in their roles, access is denied.

ikkisoft commented on rfd/0034-desktop-access-windows.md:
```diff
@@ -0,0 +1,256 @@
+`windows_desktop_service`, which is more risky (see [Security
+Concerns](#security-concerns) below).
+
+#### Agent mode
```
The overall design of agent mode introduces the risk of privilege escalation vulnerabilities since the service has to run as `system`. This is something would need to be throughly tested during the security audit.

awly replied:
Agreed :+1:

ikkisoft commented on rfd/0034-desktop-access-windows.md:
```diff
@@ -0,0 +1,256 @@
+
+RDP is a very complex protocol with a history of vulnerabilities. Using a
+memory-safe client that we fully control and not exposing raw RDP publicly is
+essential. Running Teleport in agent mode is and extra hardening step to avoid
```
s/and/an/

ikkisoft commented on rfd/0034-desktop-access-windows.md:
```diff
@@ -0,0 +1,256 @@
+### Security concerns
+
+RDP is a very complex protocol with a history of vulnerabilities. Using a
+memory-safe client that we fully control and not exposing raw RDP publicly is
```
This would have to be developed in a memory-safe language since you have to support the full RDP protocol (complex state machine!) and you might be connecting to untrusted RDP services. We should plan for extensive fuzzing on this component alone.

awly replied:
Yep, the initial attempt is using a Rust library: https://github.com/gravitational/teleport/pull/7824
To call into it from Go we need some unsafe C glue, which is a good area to focus on during review.

ikkisoft commented on rfd/0035-desktop-access-windows-authn.md:
```diff
@@ -0,0 +1,119 @@
+will use Teleport TLS certs and expose the User TLS CA for import into the
+Windows trust store.
+
+Note that RDP implements smart card authentication by proxying low-level driver
```
Similarly to the RDP client, this component would have to be implemented in a memory-safe language since it will be parsing and processing untrusted RDP server data.

ikkisoft commented on rfd/0035-desktop-access-windows-authn.md:
```diff
@@ -0,0 +1,119 @@
+state: draft
+---
+
+# RFD 35 - Desktop Access - Windows Certificate Authentication
```
I am far from being an Active Directory expert, but wouldn't this require the installation of Teleport CA in the Active Directory's NTAuth store? Have you considered the implications of that from a UX standpoint? I don't think there're better solutions, but I want to make sure that this was on your radar.

awly replied:
Correct. The only way to teach AD to trust smartcard certs is to import the CA cert into the trust store.
I have not looked at how exactly that process works yet, and it will probably be kludgy. Likely steps 1-4 of https://docs.microsoft.com/en-us/troubleshoot/windows-server/windows-security/enabling-smart-card-logon-third-party-certification-authorities#configuration-instructions

But it's the only way I found to not rely on usernames/passwords for RDP logins.

ikkisoft commented on rfd/0037-desktop-access-protocol.md:
```diff
@@ -0,0 +1,191 @@
+The protocol consists of if discrete messages, sent between a client and a
+server. These messages are passed over a secure, authenticated and reliable
+transport layer, like TLS or a websocket. The protocol leaves all security
+concerns (authentication, integrity, etc) to the transport layer.
```
We need to test for cross-origin WebSocket hijacking (see https://portswigger.net/web-security/websockets/cross-site-websocket-hijacking). Since the transport channel is fully responsible for authentication, we need to ensure that different origins in the user agent cannot tamper with the websocket channel

awly replied:
Thanks, added to https://github.com/gravitational/teleport/issues/7761

ikkisoft commented on rfd/0037-desktop-access-protocol.md:
```diff
@@ -0,0 +1,191 @@
+concerns (authentication, integrity, etc) to the transport layer.
+
+Messages are bi-directional and asynchronous - client and server can send any
+message at any time to the other end. The expected sequence of messages for a
```
> client and server can send any message at any time 

You would still need to implement a basic state machine to ensure that command control commands are not issued prior authentication.

awly replied:
Ack, the implied behavior here is dropping pre-auth messages.
Practically, it shouldn't be possible for them to succeed anyway: no messages can start until client->proxy authentication to create a websocket and no input messages can be sent on teleport->rdp server until RDP authentication succeeds.

ikkisoft commented on rfd/0037-desktop-access-protocol.md:
```diff
@@ -0,0 +1,191 @@
+Key combinations show up as a sequence of keys going into `state=1` and then
+back to `state=0`.
+
+#### 6 - clipboard data
```
What's the plan for implementing clipboard on the client (since it's a web app)? Asynchronous Clipboard API?

awly replied:
@ibeckermayer was looking into this.

ibeckermayer replied:
Yes, [Asynchronous Clipboard API](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API)

ikkisoft replied:
Thanks! 

@phosphore This would be part of the implementation review too.
